edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/DeviceScopeApiClient.cs

```diff
@@ -0,0 +1,158 @@
+
+        public Task<ScopeResult> GetIdentitiesInScope() =>
+            this.GetIdentitiesInScopeWithRetry(this.GetServiceUri(Option.None<string>()));
+
+        public Task<ScopeResult> GetNext(string continuationToken) =>
+            this.GetIdentitiesInScopeWithRetry(this.GetServiceUri(Option.Some(continuationToken)));
+
+        public Task<ScopeResult> GetIdentity(string targetDeviceId, string targetModuleId)
+        {
+            Preconditions.CheckNonWhiteSpace(targetDeviceId, nameof(targetDeviceId));            
+            return this.GetIdentitiesInScopeWithRetry(this.GetServiceUri(targetDeviceId, targetModuleId));
+        }
+
+        internal Uri GetServiceUri(Option<string> continuationToken) =>
+            continuationToken
+                .Map(c => new Uri(this.iotHubBaseHttpUri, c))
+                .GetOrElse(
+                    () =>
+                    {
+                        string relativeUri = InScopeIdentitiesUriTemplate.FormatInvariant(this.deviceId, this.moduleId, this.batchSize, null, ApiVersion);
+                        var uri = new Uri(this.iotHubBaseHttpUri, relativeUri);
+                        return uri;
+                    });
+
+        internal Uri GetServiceUri(string targetDeviceId, string targetModuleId)
+        {
+            string relativeUri = InScopeTargetIdentityUriFormat.FormatInvariant(this.deviceId, this.moduleId, targetDeviceId, targetModuleId, ApiVersion);
+            var uri = new Uri(this.iotHubBaseHttpUri, relativeUri);
+            return uri;
+        }
+
+        async Task<ScopeResult> GetIdentitiesInScopeWithRetry(Uri uri)
+        {
+            try
+            {
+                return await ExecuteWithRetry(
+                    () => this.GetIdentitiesInScope(uri),
+                    Events.RetryingGetIdentities);
+            }
+            catch (Exception e)
+            {
+                Events.ErrorInScopeResult(e);
+                throw;
+            }
+        }
+
+        async Task<ScopeResult> GetIdentitiesInScope(Uri uri)
+        {
+            var client = new HttpClient();
+            using (var msg = new HttpRequestMessage(HttpMethod.Get, uri))
+            {
+                string token = await this.edgeHubTokenProvider.GetTokenAsync(Option.None<TimeSpan>());
+                msg.Headers.Add(HttpRequestHeader.Authorization.ToString(), token);
+
+                HttpResponseMessage response = await client.SendAsync(msg);
+                string content = await response.Content.ReadAsStringAsync();
+                if (response.IsSuccessStatusCode)
+                {
+                    var scopeResult = JsonConvert.DeserializeObject<ScopeResult>(content);
+                    Events.GotValidResult();
+                    return scopeResult;
+                }
+                else
+                {
+                    throw new Exception($"Error getting Scope result from IoTHub. Status code - {response.StatusCode}, Message - {content}");
```

Comment:
Should we introduce an exception type here instead of relying on a plain `Exceptoion`?

Reply:
Oh yes.. planned to do that, forget. Let me add.